### ISSUE TITLE
Move editor from EditorJS to TipTap

### ISSUE THREAD

**Issue by deshetti** [factly/dega#454]
### Discussed in https://github.com/factly/dega/discussions/378

<div type='discussions-op-text'>

<sup>Originally posted by **deshetti** July 22, 2021</sup>
EditorJS was a really good upgrade from what we had with Quill in terms of experience and support, but it feels like that last mile improvement is not possible to make it a kickass block editor like in WordPress, Notion, Ghost etc.

I was reading a lot into folks like us experiencing the same issues with EditorJS hitting a roadblock after a while. It is still a great block editor to get started, but for folks like us who want to invest in an editor which is a huge thing for a CMS like Dega, it is getting pretty tricky to resolve some issues!

I do not want to make any assumptions or changes yet, but I really want to explore https://www.tiptap.dev/ which is built on Prosemirror. From the docs and experience from a few other folks, it looks like it is heading in the right direction. I believe this project was actively started around the same time we chose EditorJS, at that time choosing Prosemirror was definitely not going to work for us being a low-level editor and the amount of work it would have required to get a functional block editor and all the features tiptap developed on top of it.

The folks at tiptap are also building an open-source collaborative editor which seems interesting: https://www.hocuspocus.dev

Block editor is not directly supported, but we can build on the APIs they host. That is a downside to it, but not sure how much effort it would be. Based on this comment from one of the EditorJS user who moved to Prosemirror, it doesn't seem like a major challenge: https://github.com/codex-team/editor.js/issues/908#issuecomment-834347466 In this case, we would have to explore building a block editor on top of Tiptap. 

A bunch of other things that I see are well thought out and forward-looking in Tiptap are:

- Support for Markdown
  - EditorJS, the issue is open for over two years: https://github.com/codex-team/editor.js/issues/709
  - Tiptap supports writing in Markdown already: https://www.tiptap.dev/examples/default
- Keyboard shortcuts
  - EditorJS: doesn't seem to support at the lower levels: https://github.com/codex-team/editor.js/issues/1734
  - Tiptap has pretty much everything that we need currently: https://www.tiptap.dev/api/keyboard-shortcuts/ 
  - Tiptap for Markdown: https://www.tiptap.dev/examples/markdown-shortcuts/
- Support for slash commands
  - EditorJS has an open issue for around two years now: https://github.com/codex-team/editor.js/issues/927#issuecomment-608643645
  - Tiptap seems to be working on this support: https://www.tiptap.dev/guide/menus/#slash-commands-work-in-progress
- Collaboration
  - EditorJS doesn't seem to be set out to be a collaborative editor based on the current design from various comments that I read. They seem to be working on making this change and don't have a timeline: https://github.com/codex-team/editor.js/issues/1710#issuecomment-869712285
  - For Tiptap, the feature is at the core and is done using Y.js which seems like a very good implementation: https://www.tiptap.dev/api/extensions/collaboration/
- Buggy with undo/redo: https://github.com/factly/dega/issues/185
  - For EditorJS, we are currently using a plugin and it seems like this is their roadmap for a fully-featured functionality with again no tentative timelines: https://github.com/codex-team/editor.js/issues/1710#issuecomment-869712285
  - Tiptap has the functionality implemented already: https://www.tiptap.dev/api/extensions/history/#usage
- Copy/Pasting from Word. This is something @dubbudu requested as a feature.
  - For EditorJS, there is an open issue for over two years: https://github.com/codex-team/editor.js/issues/729
  - I am not sure of what the state of this is with Tiptap, will have to test it out
- Exporting content to HTML
  - For EditorJS, we had to write a custom solution in the X package
  - Looks like this is supported directly with Tiptap: https://www.tiptap.dev/guide/output/#option-2-html

I noticed that conversations for like the features above are pretty much nipped in the bud on EditorJS or just ignored forever. It is good for us to keep using EditorJS to see what all roadblocks we hit into like this and then evaluate to see if tiptap lets us gets everything done with what is there in EditorJS and more. If we ever had to migrate the editor data at a later time, we might have to write the scripts to convert the JSON from EditorJS format to Tiptap format. Or, we could explore more with the migration they have: https://www.tiptap.dev/guide/output/#migration

@shreeharsha-factly  This is maybe something that you might have to look into after VidCheck & beta versions of MandE. Nothing prior to that!! Giving it a month or two might help us as it is currently very actively developed. For eg: https://www.tiptap.dev/guide/menus/#slash-commands-work-in-progress
</div>

**Comment by ctwhome** [comment]
Hi, how did issue hold up with time? Is it something younare still considering or editorjs solved all problems you mentioned in this detailed issue?

**Comment by deshetti** [comment]
@ctwhome We decided to move to using TipTap and are building a reusable editor on top of that. We are going to release the early version of it in a couple of weeks

**Comment by ctwhome** [comment]
So you did the movement after all. How is the experience so far in comparison with editorjs?

**Comment by deshetti** [comment]
@ctwhome so far we are happy with TipTap

**Comment by ctwhome** [comment]
That's a good indication! Thanks so much for the answer 👌

**Comment by soudis** [comment]
> @ctwhome We decided to move to using TipTap and are building a reusable editor on top of that. We are going to release the early version of it in a couple of weeks

I'm at a similiar crossroads and curious if you already have an early version out? And also if you are still going down the tiptap road?

**Comment by deshetti** [comment]
@soudis Yeah, we have a very early version of the editor called [Scooter](https://github.com/factly/scooter) that we are using in a couple of applications. We are using it in an open-source CMS called [Dega](https://github.com/factly/dega) that we are developing.

This is a very early version of the editor that is continuously changing. We have some docs for it that are not updated fully as well: https://scooter.onrender.com/?path=/docs/getting-started--page

**Comment by soudis** [comment]
@deshetti thank you! Already looked into it, impressive work. I like that your implementation keeps being extendable (a lot of other editors on top of tiptap don't).

**Comment by adamtaylor13** [comment]
@deshetti Same consensus here in The Future™? We're making a decision between both, and your list of pros/cons at the top was very helpful. I'd love some 2-years-in-hindsight commentary if you'd care to expound.

**Comment by deshetti** [comment]
> @deshetti Same consensus here in The Future™? We're making a decision between both, and your list of pros/cons at the top was very helpful. I'd love some 2-years-in-hindsight commentary if you'd care to expound.

@adamtaylor13 Sorry for the delayed response. We have developed our own editor called Scooter on TipTap & ProseMirror that is open source. We have made quite a few updates to the editor in the past few months including adding AI features similar to Notion: https://github.com/factly/scooter

While the latest code is available in the above repo, we haven't got a chance to update the documentation site or modify the storybook components to use the latest version yet. We are testing the editor internally at Factly in a couple of applications and are going to make to make a general announcement for an alpha release by the end of the month including a detailed documentation site.

As a general commentary about TipTap, it is definitely a step up from where we were with EditorJS. We ran into some issues in TipTap that we were tricky to fix and some for eg like [this](https://github.com/ueberdosis/tiptap/issues/4063)which is still pending with no comments from anyone. But, overall we are happy with the shift to TipTap.
